Fixed our tests and improved the code coverage

bpoplauschi opened the pull request:
### New Pull Request Checklist

* [X] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [X] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [X] I have searched for a similar pull request in the [project](https://github.com/rs/SDWebImage/pulls) and found none

* [X] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [X] I have added the required tests to prove the fix/feature I am adding
* [X] I have updated the documentation (if necessary)
* [X] I have run the tests and they pass
* [X] I have run the lint and it passes (`pod lib lint`)

### Fixed our tests for good and improved code coverage
- Added explicit tests for `SDWebImageGIFCoder`
- Refactored the `SDWebImageDecoderTests` a bit. Per DRY, created `verifyCoder:withLocalImageURL:isAnimatedImage:` method that will take a imageURL and a coder and will check decoding and encoding back and forth, sizes and so. Reusing this method for the static WebP image, for the animated WebP image and for the GIF.
- Switched all the tests from using https://s3.amazonaws.com/fast-image-cache/demo-images/* images which are big and sometimes fail to arrive in 5 seconds (before the test can timeout) to smaller images, basically using http://via.placeholder.com/ like http://via.placeholder.com/50x50.jpg or http://via.placeholder.com/200x200.png
This should speed up the tests a bit and also avoid some of the random failures due to timeout
- Finally fixed the tests from `SDImageCacheTests` so we can re-enable `test40InsertionOfImageData`
  - the problems: some tests were just poorly written, using async calls but not properly setting async expectations. Those were generating undefined behaviour (not clear if/when those callbacks would be executed).
  - also, we wanted to call `clearAllCaches` before every test, but since some methods there are async, this doesn't work with the way `setUp` was designed. So basically sometimes before a test we could clear the cache but sometimes not, or even the clear could come while running the test which was worst
  - decided to fix this by making sure after every test that we manually clear what we have added.
- Updated the `Tests` `Podfile` so we don't get a warning (`xcodeproj` was renamed to `project`)

bpoplauschi commented:
I created a dedicated PR so I can check codecov, since this affects coverage (should show an increase from aprox 74% to 77%).

bpoplauschi commented on Tests/Tests/SDImageCacheTests.m:
```diff
@@ -46,11 +46,15 @@ - (void)test04ClearDiskCache{
             if (!isInCache) {
-                [expectation fulfill];
+                [self.sharedImageCache calculateSizeWithCompletionBlock:^(NSUInteger fileCount, NSUInteger totalSize) {
+                    expect(fileCount).to.beLessThan(100);
```
actually, this should be check against 0, because after clearDisk we should have 0 files in the cache.

bpoplauschi commented:
Looks good :) @dreampiggy @skyline75489 merge it when you can

dreampiggy commented on Tests/Tests/SDWebImageDecoderTests.m:
```diff
@@ -120,8 +126,36 @@ - (void)test10ThatAnimatedWebPCoderWorks {
+    expect([[SDWebImageGIFCoder sharedCoder] canEncodeToFormat:SDImageFormatGIF]).to.beTruthy();
+    expect([[SDWebImageImageIOCoder sharedCoder] canEncodeToFormat:SDImageFormatGIF]).to.beTruthy();
+    NSData *outputData = [[SDWebImageGIFCoder sharedCoder] encodedDataWithImage:gifImage format:SDImageFormatGIF];
+    expect(outputData).toNot.beNil();
```
Sorry...I forget this task. By the way, what about decode this `outputData` again and check frame count, image size and so on? This can ensure that it works for both decoding & encoding process :D

bpoplauschi replied:
@dreampiggy no prob. I can take care of this. I just copy-pasted from the WebP tests. I will revisit those as well and also fix my comment above (about that `calculateSize` expectancy).

bpoplauschi replied:
Done @dreampiggy.
Per DRY principle, created `verifyCoder:withLocalImageURL:isAnimatedImage:` method that will take a imageURL and a coder and will check decoding and encoding back and forth, sizes and so. Reusing this method for the static WebP image, for the animated WebP image and for the GIF.

bpoplauschi commented:
`test02PrefetchMultipleImages` failed during the last build. I think this is just a random failure, I will re-trigger.

bpoplauschi commented:
Guys, a few more commits.
I am happy with the current status of our tests:
- Refactored the `SDWebImageDecoderTests` a bit. Per DRY, created `verifyCoder:withLocalImageURL:isAnimatedImage:` method that will take a imageURL and a coder and will check decoding and encoding back and forth, sizes and so. Reusing this method for the static WebP image, for the animated WebP image and for the GIF.
- Switched all the tests from using https://s3.amazonaws.com/fast-image-cache/demo-images/* images which are big and sometimes fail to arrive in 5 seconds (before the test can timeout) to smaller images, basically using http://via.placeholder.com/ like http://via.placeholder.com/50x50.jpg or http://via.placeholder.com/200x200.png
This should speed up the tests a bit and also avoid some of the random failures due to timeout
- Finally fixed the tests from `SDImageCacheTests` so we can re-enable `test40InsertionOfImageData`
  - the problems: some tests were just poorly written, using async calls but not properly setting async expectations. Those were generating undefined behaviour (not clear if/when those callbacks would be executed).
  - also, we wanted to call `clearAllCaches` before every test, but since some methods there are async, this doesn't work with the way `setUp` was designed. So basically sometimes before a test we could clear the cache but sometimes not, or even the clear could come while running the test which was worst
  - decided to fix this by making sure after every test that we manually clear what we have added.

bpoplauschi commented:
Finally, green light from CI. This PR is ready.

bpoplauschi commented:
@dreampiggy merge it please :)
